Examples of matrix input usage in Spatial Shaders

paddy-exe opened the pull request:
Introduces examples for matrix input usage to users in Spatial Shaders such as:
* object's world space position
* camera's world space position
* camera eye vector in world space direction of the camera
* camera's view direction to the object in view space

clayjohn commented on tutorials/shaders/shader_reference/spatial_shader.rst:
```diff
@@ -141,6 +141,19 @@ Users can override the modelview and projection transforms using the ``POSITION`
+
+.. note::
+
+    MODELVIEW_MATRIX combines both the the MODEL_MATRIX and VIEW_MATRIX and is better suited when floating point issues may arise
```
```suggestion
    ``MODELVIEW_MATRIX`` combines both the the ``MODEL_MATRIX`` and ``VIEW_MATRIX`` and is better suited when floating point issues may arise. For example, if the object is very far away from the world origin, you may run into floating point issues when using the separated ``MODEL_MATRIX`` and ``VIEW_MATRIX``.
```

clayjohn replied:
This suggestion is still relevant.

clayjohn commented on tutorials/shaders/shader_reference/spatial_shader.rst:
```diff
@@ -225,6 +238,20 @@ The default use of a Godot fragment processor function is to set up the material
+
+.. note::
+
+    A commonly used alternative to the ``matrix[3].xyz`` is to use ``vec3 origin = (MODEL_MATRIX * vec4(0,0,0,1)).xyz`` 
```
```suggestion
    A commonly used alternative to the ``matrix[3].xyz`` is to use ``vec3 origin = (MODEL_MATRIX * vec4(0,0,0,1)).xyz``. 
```

clayjohn commented on tutorials/shaders/shader_reference/spatial_shader.rst:
```diff
@@ -141,6 +141,19 @@ Users can override the modelview and projection transforms using the ``POSITION`
+    
+    render_mode skip_vertex_transform;
+    void vertex() {
+        vec3 camera_direction  = (MODELVIEW_MATRIX * vec4(VERTEX, 1.0)).xyz;
```
This is just vertex position in Camera space (which is equivalent to camera direction to vertex), it is not camera direction to object. Camera direction to object is typically ``(MODELVIEW_MATRIX * vec4(0.0, 0.0, 0.0 1.0)).xyz;`` when the mesh is centered about the origin. 

I don't know if this is a really helpful example.

paddy-exe replied:
My original code was actually this: `normalize(WORLD_MATRIX[3].xyz - CAMERA_MATRIX[3].xyz) = Camera direction to object in world space` but Lyuma mentioned that it would be confusing since it would be too specific for a billboard shader

clayjohn replied:
I think it fits in fine with the other examples below. As long as they are framed with "here is how you can calculate some common variables using just the built ins"

clayjohn commented on tutorials/shaders/shader_reference/spatial_shader.rst:
```diff
@@ -225,6 +238,20 @@ The default use of a Godot fragment processor function is to set up the material
 
+Below are examples of frequently used matrix inputs for shaders.
+
+.. code-block:: glsl
+
+    MODEL_MATRIX[3].xyz = Object's world space position // global_transform.origin in GDScript
+    // Using mat3(MODEL_MATRIX) is the same as the global_transform.basis in GDScript
+    INV_VIEW_MATRIX[3].xyz = Camera's world space position // camera.global_transform.origin in GDScript
+    VIEW_MATRIX[3].xyz = Camera eye vector in world space direction of the camera
```
I wonder if this makes more sense in a table or some other format. Something like:

Many helpful values can be found within the Matrix built-ins themselves, here are a few:

| Built In                                              | Description                                                                                      |
|-------------------------------------------------------|--------------------------------------------------------------------------------------------------|
| ``MODEL_MATRIX[3].xyz``                               | Object's world space position. It is also equivalent to ``global_transform.origin`` in GDScript. |
| ``mat3(MODEL_MATRIX)``                                | etc.                                                                                             |
| ``INV_VIEW_MATRIX[3].xyz``                            | etc.                                                                                             |
| ``VIEW_MATRIX[3].xyz``                                | etc.                                                                                             |
| ``(MODELVIEW_MATRIX * vec4(0.0, 0.0, 0.0, 1.0)).xyz`` | etc.                                                                                             |

clayjohn replied:
Fair enough, I don't have a strong opinion one way or the other. I don't love the blend between code and text. So perhaps, it would make sense to do more psuedocode with a comment like in the shading reference.

e.g.: 
```
vec3 world_space_position = MODEL_MATRIX[3].xyz; // Object's world space position. This is also equivalent to global_transform.origin in GDScript.
```

paddy-exe commented:
@clayjohn I incorporated your feedback in and moved the code together directly above the `Fragment built-ins`. I did leave the note about the `MODEL_VIEW_MATRIX` there though since it seems quite useful to know. Let me know what you think about this.

clayjohn commented on tutorials/shaders/shader_reference/spatial_shader.rst:
```diff
@@ -218,13 +218,32 @@ shader, this value can be used as desired.
+.. code-block:: glsl
+
+    vec3 model_world_space = MODEL_MATRIX[3].xyz // Object's world space position. This is the equivalent to global_transform.origin in GDScript.
+    // Using mat3(MODEL_MATRIX) is the same as the global_transform.basis in GDScript.
```
This deserves its own example the same as the others. It looks out of place to have this one totally in a comment while the others have examples.

paddy-exe replied:
Done!

clayjohn commented on tutorials/shaders/shader_reference/spatial_shader.rst:
```diff
@@ -218,13 +218,32 @@ shader, this value can be used as desired.
 
+.. note::
+
+    MODELVIEW_MATRIX combines both the the MODEL_MATRIX and VIEW_MATRIX and is better suited when floating point issues may arise
```
```suggestion
    ``MODELVIEW_MATRIX`` combines both the the ``MODEL_MATRIX`` and ``VIEW_MATRIX`` and is better suited when floating point issues may arise. For example, if the object is very far away from the world origin, you may run into floating point issues when using the separated ``MODEL_MATRIX`` and ``VIEW_MATRIX``.
```

Copying this over from my previous review comment as it appears to have been missed

clayjohn commented on tutorials/shaders/shader_reference/spatial_shader.rst:
```diff
@@ -218,13 +218,32 @@ shader, this value can be used as desired.
+
+.. note::
+
+    A commonly used alternative to ``matrix[3].xyz`` is to use ``vec3 origin = (MODEL_MATRIX * vec4(0,0,0,1)).xyz`` 
```
```suggestion
    A commonly used alternative to ``MODEL_MATRIX[3].xyz`` is to use ``vec3 origin = (MODEL_MATRIX * vec4(0,0,0,1)).xyz``. It is more efficient to use ``MODEL_MATRIX[3].xyz`` as it avoids the matrix multiplication.
```

clayjohn reviewed (approved):
Looks great! Thanks!
